### PR TITLE
(tree) Added runTransaction API to tree view

### PR DESCRIPTION
**Pull request by agarwal-navin**
## Description
Added an alpha `runTransaction` API to tree view (TreeViewAlpha):
- It takes an object as parameter. It currently has the following paramters which can be extended in the future:
  - A `transaction` callback param which is function to run as the body of the transaction.
  - A `preconditions` param which can used to specify constraints which should be satisfied before the transaction is run.
- It reutrns an object as a result. The result tells whether the transaction was a success or failure along with an optional user defined value. It can be extended to add other properties.
There are 2 main variations of the API - One where the user can declare a return value type that will be returned by both the `transaction` delegate and the outer `runTransaction` API. The other where it simply returns whether the transaction was sucessful or failures and an empty return indicates success.

The rest of the implementation is the same as the standalone `runTransaction` API.

[AB#25741](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/25741)

### CODE REVIEW

File: packages/dds/tree/src/shared-tree/transactionTypes.ts [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+/**
+ * The status of a transaction on the tree view.
+ * @alpha
+ */
+export interface TransactionStatus<TResult> {
+	/**
+	 * The value returned by the inner `transaction` function or the special {@link RunTransaction.rollback | rollback value}
+	 * (`Tree.runTransaction.rollback`) which means that the transaction was aborted.
+	 */
+	readonly result: TResult | typeof rollback;
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/shared-tree/transactionTypes.ts:65]: I'm wondering if we want to switch to a pattern where the return value is orthogonal to whether the transaction is rolled rolled back. This would be useful to app authors that want to return some result in the rollback case.

@noencke, thoughts on that?

File: packages/dds/tree/src/shared-tree/transactionTypes.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+ * Parameters for running a transaction on the tree view that applies one or more edits to the tree as a single atomic unit.
+ * @alpha
+ */
+export interface TransactionParams<TResult> {
+	/**
+	 * The function to run as the body of the transaction. This function is passed the root of the tree.
+	 * At any point during the transaction, the function may return the special {@link RunTransaction.rollback | rollback value}
+	 * (`Tree.runTransaction.rollback`) to abort the transaction and discard any changes it made so far.
+	 */
+	readonly transaction: () => TransactionStatus<TResult>;
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/shared-tree/transactionTypes.ts:46]: I wonder if we should make this
```suggestion
	readonly transaction: () => TransactionStatus<TResult> | undefined;
```
so that transaction authors don't have to return anything if they don't want to.

File: packages/dds/tree/src/test/shared-tree/treeNodeApi.spec.ts [hunk 3]
```diff
@@ -56,59 +56,360 @@ describe("treeApi", () => {
...
+				view.runTransaction({
+					transaction: () => {
+						view.root.content = 43;
+						return { result: Tree.runTransaction.rollback };
+					},
 				});
 				assert.equal(view.root.content, 42);
 			});
 
 			it("rolls back transactions on error", () => {
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/test/shared-tree/treeNodeApi.spec.ts:399]:
```suggestion
			it("breaks the view on error", () => {
```

File: packages/dds/tree/src/shared-tree/transactionTypes.ts (hunk 1 above)
Comment (noencke) [reply packages/dds/tree/src/shared-tree/transactionTypes.ts:65]: Yes - I was envisioning something like:

```ts
TransactionResult<TResult = unknown> {
  readonly status: "committed" | "aborted";
  readonly result?: TResult;
}
```

File: packages/dds/tree/src/shared-tree/transactionTypes.ts (hunk 2 above)
Comment (noencke) [reply packages/dds/tree/src/shared-tree/transactionTypes.ts:46]: I think that's a good idea. I don't see any reason to prevent users from having a "fast path" for simple transactions that don't want to bother with a return value and are always successful (no return => committed/success).

Ideally I think we'd want something like `TransactionResult<TResult> | void`, rather than `undefined`. But, I'm not sure that is correct - because a function can return _anything_ and still satisfy a return value of `void`. So, we'd have to use a type guard to guess whether the return value is a transaction result, which isn't guaranteed to never collide with some other user type that happen to be returned from their drop-in transaction function (fringe and unlikely scenario, I know...).

Following that logic, I think I see three reasonable possibilities:

1. `() => TransactionStatus<TResult> | "committed" | "aborted"`
  If we're going to let them return `undefined` we may as well have them return a status signal instead. However, if they want to pass up a return value then they still need to use the `TransactionStatus`
2. `() => TransactionStatus<TResult> | TResult`
  This makes returning things easier, but allows no shortcut for abort/rollback and has the drawback mentioned above (`TResult` might be some object that looks like a `TransactionStatus` erroneously).

I think I want to get more second opinions on this one from API Council and Nick. It's not clear to me what the best option is. There are other possibilities too - e.g. we could have functions available that make it easy to construct the TransactionStatus object. So:

`return commit()`
`return abort()` (or `return rollback()`)
`return commit(myResult)`
`return abort(myResult)`

But then, where do those functions come from? Are they imported free functions? Or can we hang them off of something relevant that we know the user already has - like we did with the `rollback` symbol?

File: packages/dds/tree/src/shared-tree/transactionTypes.ts [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+/**
+ * Parameters for running a transaction on the tree view that applies one or more edits to the tree as a single atomic unit.
+ * @alpha
+ */
+export interface TransactionParams<TResult> {
+	/**
+	 * The function to run as the body of the transaction. This function is passed the root of the tree.
+	 * At any point during the transaction, the function may return the special {@link RunTransaction.rollback | rollback value}
+	 * (`Tree.runTransaction.rollback`) to abort the transaction and discard any changes it made so far.
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/shared-tree/transactionTypes.ts:44]: These comments will need updating once we settle the API

File: packages/dds/tree/src/shared-tree/transactionTypes.ts [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+	 * this client and ignored by all other clients.
+	 */
+	readonly preconditions?: readonly TransactionConstraint[];
+}
+
+/**
+ * The status of a transaction on the tree view.
+ * @alpha
+ */
+export interface TransactionStatus<TResult> {
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/shared-tree/transactionTypes.ts:60]: I think the "Status" part of the name doesn't fit well for something that is returned to us by the app code> I think of "status" as something you query. What do you guys think of `TransactionConclusion` or `TransactionEnd`? I don't _love_ them either TBH.

(note: `TransactionResult` is already taken)

File: packages/dds/tree/src/shared-tree/schematizingTreeView.ts [hunk 6]
```diff
@@ -219,6 +229,44 @@ export class SchematizingSimpleTreeView<
 		return this.view;
 	}
 
+	/**
+	 * Run a transaction which applies one or more edits to the tree as a single atomic unit.
+	 */
+	public runTransaction<TResult>(
+		params: TransactionParams<TResult>,
+	): TransactionStatus<TResult> {
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/shared-tree/schematizingTreeView.ts:237]: I think we may want to decouple the type that is returned by `TransactionParams.transaction()` from the type returned by `runTransaction`.

For example, I don't see much value in including the constraints in the object that is returned by `runTransaction`.
In the future, we'll likely also want to add a bunch of useful things to the object that is returned by `runTransaction` (like something to keep track of whether the associated commit has been sequenced) and that's definitely not something that the user could supply in the object they return from `TransactionParams.transaction()`.

File: packages/dds/tree/src/shared-tree/transactionTypes.ts (hunk 5 above)
Comment (noencke) [reply packages/dds/tree/src/shared-tree/transactionTypes.ts:60]: (`TransactionResult` is taken but it's not exported, and it's not really used for a whole lot. If that ends up being the name we like for this purpose, we should steal it)

File: packages/dds/tree/src/shared-tree/transactionTypes.ts (hunk 2 above)
Comment (yann-achard-MS) [reply packages/dds/tree/src/shared-tree/transactionTypes.ts:46]: Note that making the return type `TransactionResult<TResult> | undefined` doesn't force them to return anything but it does ensure that they don't return something that isn't a `TransactionResult<TResult>` or `undefined`.

File: packages/dds/tree/src/shared-tree/transactionTypes.ts [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+/**
+ * The status of a transaction on the tree view.
+ * @alpha
+ */
+export interface RunTransactionResult<TResult> {
+	/**
+	 * The value returned by the inner `transaction` function or the special {@link RunTransaction.rollback | rollback value}
+	 * (`Tree.runTransaction.rollback`) which means that the transaction was aborted.
+	 */
+	readonly result: TResult | typeof rollback | undefined;
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/shared-tree/transactionTypes.ts:83]: What is the motivation for the ` | undefined` here?
I might be missing something, but I think having it here allows runTransaction to discard the TResult returned by the `transaction` delegate instead of returning it, which would be a bug.

File: packages/dds/tree/src/shared-tree/transactionTypes.ts [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+	 * it will be rolled back on this client and ignored by all other clients.
+	 */
+	readonly undoPreconditions?: readonly TransactionConstraint[];
+}
+/**
+ * The failed outcome of a transaction, i.e. the transaction should abort.
+ * @alpha
+ */
+export interface AbortTransaction {
+	/** The failed outcome indicating the rest of the transaction should abort and any changes should be rolled back */
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/shared-tree/transactionTypes.ts:57]: Would be good to specify that this only aborts the current transaction scope (not the outer scopes, if any)

File: packages/dds/tree/src/shared-tree/transactionTypes.ts [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+ * @remarks The node must be in the document (its {@link TreeStatus | status} must be {@link TreeStatus.InDocument | InDocument}) to qualify as "existing".
+ * @public
+ */
+export interface NodeInDocumentConstraint {
+	readonly type: "nodeInDocument";
+	readonly node: TreeNode;
+}
+
+/**
+ * The successful outcome of a transaction, i.e. the transaction should continue.
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/shared-tree/transactionTypes.ts:37]: If we keep this "continue" name, we should clarify here that when this is returned from the outer-most transaction scope, the transaction will end.

File: packages/dds/tree/src/shared-tree/transactionTypes.ts [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+	 * @returns The result of the transaction. The user provided result (TResult) can either be returned directly or
+	 * as part of the `TransactionOutcome` object which can include other properties.
+	 * It could return nothing (TResult == void) to indicate a successful transaction.
+	 *
+	 * At any point during the transaction, the function may return the special {@link RunTransaction.rollback | rollback value}
+	 * (`Tree.runTransaction.rollback`) to abort the transaction and discard any changes it made so far.
+	 */
+	readonly transaction: () =>
+		| ContinueTransactionExt<TSuccessValue>
+		| AbortTransactionExt<TFailureValue>;
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/shared-tree/transactionTypes.ts:152]: We'd be supporting `ContinueTransaction | AbortTransaction` and `ContinueTransactionExt<TSuccessValue> | AbortTransactionExt<TFailureValue>` but not `ContinueTransaction | AbortTransactionExt<TFailureValue>` or `ContinueTransactionExt<TSuccessValue> | AbortTransaction`.

I'm cool with that, but I wanted to call it out, to make sure this is a conscious choice.

File: packages/dds/tree/src/shared-tree/transactionTypes.ts (hunk 9 above)
Comment (agarwal-navin) [reply packages/dds/tree/src/shared-tree/transactionTypes.ts:37]: Not sure I follow. I used the word "continue" to convey from the user's perspective that the user part or body of the transaction was successful and the system part of the transaction process can continue. But I guess that's a little confusing when it comes to nested transaction.
Maybe we can use some other word. I thought of "run" but that somehow doesn't convey the right idea. Also, it has to be apposite of "abort". Maybe "confirm" or "complete"?

File: packages/dds/tree/src/shared-tree/transactionTypes.ts (hunk 10 above)
Comment (agarwal-navin) [reply packages/dds/tree/src/shared-tree/transactionTypes.ts:152]: I am fine with that too. We can always add it if needed. We could still support it because the implementation wouldn't change. However, it will add more types and make it complex.

File: packages/dds/tree/src/shared-tree/transactionTypes.ts (hunk 7 above)
Comment (agarwal-navin) [reply packages/dds/tree/src/shared-tree/transactionTypes.ts]: This has changed in the latest iteration.

File: packages/dds/tree/src/shared-tree/schematizingTreeView.ts [hunk 11]
```diff
@@ -213,6 +226,58 @@ export class SchematizingSimpleTreeView<
 		return this.view;
 	}
 
+	/**
+	 * Run a transaction which applies one or more edits to the tree as a single atomic unit.
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/shared-tree/schematizingTreeView.ts:230]: I think both overloads need doc comments.

You may want to look at the doc comments on the existing API for inspiration. I think the important parts are "Apply one or more edits to the tree as a single atomic unit" and "All of the changes in the transaction are applied synchronously and therefore no other changes (either from this client or from a remote client) can be interleaved with those changes. Note that this is guaranteed by Fluid for any sequence of changes that are submitted synchronously, whether in a transaction or not. However, using a transaction has the following additional consequences: - If reverted (e.g. via an "undo" operation), all the changes in the transaction are reverted together. - The internal data representation of a transaction with many changes is generally smaller and more efficient than that of the changes when separate."

Comment (yann-achard-MS) [reply packages/dds/tree/src/shared-tree/schematizingTreeView.ts:230]: I guess it would also be good to say something about the fact that it is okay to call this within a transaction, and explain what that means (abort only aborts the subtransaction, constraints apply to the outer transaction, undo will undo the outer transaction).

File: packages/dds/tree/src/shared-tree/transactionTypes.ts (hunk 9 above)
Comment (yann-achard-MS) [reply packages/dds/tree/src/shared-tree/transactionTypes.ts:37]: Ah, I can see how "continue" makes sense in that light. What I find the most confusing about it is that I would expect it to mean "please keep the transaction open for more changes that will add later (either in this JS turn or a later one)" which is something we used to support and might support again.

It's surprisingly hard to find something good here.
* "commit" seems odd when it comes to nested transactions. It also is confusing when you think about the fact that changes are only truly committed once they've been sequenced (but perhaps we can say the changes are committed but not pushed)
* "apply" seems wrong because we apply changes as we go so all changes are already applied.
* "ok" is not bad.

I think my preference might be to just not have a `result` field on this type or at least make sure that the user doesn't have to populate it (we may still want to field declared for the sake of using the type as a discriminated union).

Otherwise, "commit" seems okay. Either way I'm leaning for the type to be named `CommitTransaction`.

File: packages/dds/tree/src/shared-tree/transactionTypes.ts [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+export interface RunTransactionFailedExt<TFailureValue> extends RunTransactionFailed {
+	/** The user defined return value on failing the transaction */
+	readonly returnValue: TFailureValue;
+}
+
+/**
+ * Parameters for running a transaction on the tree view that applies one or more edits to the tree as a single atomic unit.
+ * @alpha
+ */
+export interface RunTransactionParams {
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/shared-tree/transactionTypes.ts:119]: It's dawning on me that we may be able to add a `RunSimpleTransactionParams` variant whose `transaction` field is typed as `() => void`. This would allow us to keep supporting transactions that never want to return anything (which seems very common).

File: packages/dds/tree/src/shared-tree/transactionTypes.ts [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+	 * @returns The result of the transaction indicating whether the transaction should continue or abort.
+	 */
+	readonly transaction: () => ContinueTransaction | AbortTransaction;
+	/**
+	 * An optional list of {@link TransactionConstraint | constraints} that are checked just before the transaction begins.
+	 * If any of the constraints are not met when `runTransaction` is called, it will throw an error.
+	 * If any of the constraints are not met after the transaction has been ordered by the service, it will be rolled back on
+	 * this client and ignored by all other clients.
+	 */
+	readonly preconditions?: readonly TransactionConstraint[];
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/shared-tree/transactionTypes.ts:131]: The fact that we have two variants of `RunTransactionParams` (maybe three, see comment above) makes me want to put `preconditions` in on a `BaseRunTransactionParams` interface (final name TBD) that all the `RunTransactionParams` extend.

The alternative would be to extract the `transaction` parameter into its own parameter. This would lead to: `runTransaction(transaction: () => ContinueTransaction | AbortTransaction, params?: RunTransactionParams): RunTransactionSucceeded | RunTransactionFailed`
That would reduce bring the number of `RunTransactionParams` types back down to one. I think this is a pretty common pattern: use mandatory arguments for the things that must be provided, and use a single object for the optional config stuff.

File: packages/dds/tree/src/shared-tree/transactionTypes.ts [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+	  }
+) & {
+	// Todo: For demonstration only. To be removed.
+	undoPreconditions?: readonly TransactionConstraint[];
+};
+
+/**
+ * The status of a the transaction callback in the {@link RunTransaction | RunTransaction} API. This is the
+ * same as {@link TransactionCallbackStatusExt} but with the `value` field omitted. This is useful when
+ * the user doesn't want to return any value from the transaction callback.
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/shared-tree/transactionTypes.ts:61]: who "the user" is tends to shift depending on who is reading/speaking: app authors think of "the user" as the end user interacting with the app at runtime.
How about:
```suggestion
 * the transaction callback doesn't need to return a value.
```

File: packages/dds/tree/src/shared-tree/transactionTypes.ts [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+ * @alpha
+ */
+export type TransactionResultExt<TSuccessValue, TFailureValue> =
+	| TransactionResultSuccess<TSuccessValue>
+	| TransactionResultFailed<TFailureValue>;
+
+/**
+ * The result of the {@link RunTransaction | RunTransaction} API. This is the same as {@link TransactionResultExt}
+ * but with the `value` field omitted. This is useful when the user doesn't want to return any value from the
+ * transaction.
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/shared-tree/transactionTypes.ts:104]:
```suggestion
 * but with the `value` field omitted. This is useful when the transaction callback doesn't need to return a value.
```

File: packages/dds/tree/src/test/shared-tree/treeNodeApi.spec.ts [hunk 16]
```diff
@@ -56,59 +56,480 @@ describe("treeApi", () => {
...
+		describe("view.runTransaction:transaction() return values", () => {
+			it("implicit success, no return value", () => {
 				const view = getTestObjectView();
-				run(view, (root: TestObject) => {
-					assert.equal(root, view.root);
+				const runTransactionResult = view.runTransaction(() => {
+					view.root.content = 43;
 				});
+				assert.equal(view.root.content, 43);
+				assert.equal(runTransactionResult.success, true);
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/test/shared-tree/treeNodeApi.spec.ts:383]: For all of these tests, I propose doing an asset.deepEqual on the whole `runTransactionResult` object to get more coverage. This will for example ensure that we don't put a bogus `value` in the returned object.

File: packages/dds/tree/src/test/shared-tree/treeNodeApi.spec.ts [hunk 17]
```diff
@@ -56,59 +56,480 @@ describe("treeApi", () => {
...
+						value: view.root.content,
+						undoPreconditions: [{ type: "nodeInDocument", node: view.root }],
+					};
+				});
+				assert.equal(runTransactionResult.success, false);
+				// Note that this is the value that was set before the transaction was rolled back.
+				assert.equal(view.root.content, 42);
+				// After the transaction is rolled back, the value is reverted to the original value.
+				assert.equal(runTransactionResult.value, 43);
+			});
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/test/shared-tree/treeNodeApi.spec.ts:484]: I don't think we need to test every permutation of return fields. So long as each return field is tested, I think we can remove these two tests.

File: packages/dds/tree/src/simple-tree/api/tree.ts [hunk 18]
```diff
@@ -542,6 +548,20 @@ export interface TreeViewAlpha<
 
 	// Override the base branch method to return a typed view rather than merely a branch.
 	fork(): ReturnType<TreeBranch["fork"]> & TreeViewAlpha<TSchema>;
+
+	/**
+	 * Run a transaction which applies one or more edits to the tree as a single atomic unit.
+	 * @param transaction - The transaction callback to run.
+	 * @param params - The optional parameters for the transaction.
+	 */
+	runTransaction<TSuccessValue, TFailureValue>(
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/simple-tree/api/tree.ts:557]: Both of these overloads need a more complete doc comment.

File: packages/dds/tree/src/test/shared-tree/treeNodeApi.spec.ts [hunk 19]
```diff
@@ -56,59 +56,480 @@ describe("treeApi", () => {
...
+					// each time (detach and attach passes).
+					assert.equal(shallowEventCount, 2);
+					assert.equal(deepEventCount, 2);
+				});
+
+				// TODO: When SchematizingSimpleTreeView supports forking, add test coverage to ensure that transactions work properly on forks
+			});
+		});
+
+		describe("view.runTransaction:transaction() return values", () => {
```
Comment (noencke) [inline packages/dds/tree/src/test/shared-tree/treeNodeApi.spec.ts:376]: I hate to be that guy... but technically all these new tests should go in `schematizingSimpleTreeView.spec.ts`, not in `treeNodeApi.spec.ts`, since `view.runTransaction` has nothing to do with the static `TreeNodeAPI`. If it's a huge pain to move it right now then I'm okay with it. But probably the best eventual state is to have the "common" transaction tests live in a helper in their own file, which gets consumed by TreeNodeAPI tests and view tests separately.

File: packages/dds/tree/src/test/shared-tree/treeNodeApi.spec.ts [hunk 20]
```diff
@@ -56,59 +56,480 @@ describe("treeApi", () => {
...
-				run(view, (root: TestObject) => {
-					assert.equal(root, view.root);
+				const runTransactionResult = view.runTransaction(() => {
+					view.root.content = 43;
 				});
+				assert.equal(view.root.content, 43);
+				assert.equal(runTransactionResult.success, true);
+			});
+
+			it("explicit success with return value", () => {
```
Comment (noencke) [inline packages/dds/tree/src/test/shared-tree/treeNodeApi.spec.ts:386]: I would just call this "explicit success" - I thought that by "return value" you meant "user-defined value", but you just mean "the way we express explicit success is via the returned object"

File: packages/dds/tree/src/test/shared-tree/treeNodeApi.spec.ts [hunk 21]
```diff
@@ -56,59 +56,480 @@ describe("treeApi", () => {
...
+					};
+				});
+				assert.equal(runTransactionResult.success, false);
+				// Note that this is the value that was set before the transaction was rolled back.
+				assert.equal(view.root.content, 42);
+				// After the transaction is rolled back, the value is reverted to the original value.
+				assert.equal(runTransactionResult.value, 43);
+			});
+
+			it("void return weirdness", () => {
```
Comment (noencke) [inline packages/dds/tree/src/test/shared-tree/treeNodeApi.spec.ts:486]: I don't think this is testing anything that is unique to `void` or even to this API. Unless I'm misunderstanding something, [this is just how TS works in general](https://www.typescriptlang.org/play/?#code/JYOwLgpgTgZghgYwgAgJLIN4FgBQz-IBucANgK4QBcyIZAtgEbQDcuAvrrjGSAmMAHsQyGAAoYIaqICUyALwA+NNOqEBwACaYOOTjjEz5S7HgIB6M8gCiADzBQ4yAA5QBT6GACeuAsigQwMihhDCJSCmoAZgAaMPIIACZqABZkNlYcNmkMvVwLNGEAcwgQaFJYgBUAZWQNAQgAZxAAcjBnf0IStrgNDWB+IVJ2tw9gRpFXOmQmUELkIXmGACsIPgaAOjzLCoALFCEST2R+OhR+2vqG453gK9vkAHc94U8BMmb-WtWSOChZ5BIAgQQ2IfzgDBI4zAOzgbQAVHDXmRkDDOsgGvYhIVDgBaLzuLRIqANCAkGDIUDXFAab6-WGCEAIyh6BBCDEU6joOSYHwEYjxKLRXn4fkUJLIZLsPT5Xa3WILaH7RVQFFwEAaWL3GCgKgstltYAJeQ80wi8JUZAxYVxCjJcWSzI5HCskDs4CRTnGw3MIA). So I'd just remove it.

File: packages/dds/tree/src/shared-tree/transactionTypes.ts [hunk 22]
```diff
@@ -0,0 +1,123 @@
...
+ * The result of the {@link RunTransaction | RunTransaction} API when it was successful.
+ * @alpha
+ */
+export interface TransactionResultSuccess<TSuccessValue> {
+	/** Indicates that the transaction was successful. */
+	success: true;
+	/** The user defined value when the transaction was successful. */
+	value: TSuccessValue;
+	// Todo: For demonstration only. To be removed.
+	commitId?: string;
```
Comment (noencke) [inline packages/dds/tree/src/shared-tree/transactionTypes.ts:79]: Just putting a comment here so we don't forget to remove :)

File: packages/dds/tree/src/test/shared-tree/treeNodeApi.spec.ts (hunk 19 above)
Comment (agarwal-navin) [reply packages/dds/tree/src/test/shared-tree/treeNodeApi.spec.ts:376]: Good point. Agreed that it belongs to that test file. I will try and move them.

File: packages/dds/tree/src/test/shared-tree/treeNodeApi.spec.ts (hunk 21 above)
Comment (agarwal-navin) [reply packages/dds/tree/src/test/shared-tree/treeNodeApi.spec.ts:486]: I put it there for review purposes. Agreed that it doesn't add any value. Will remove it.

File: packages/dds/tree/src/shared-tree/transactionTypes.ts (hunk 22 above)
Comment (agarwal-navin) [reply packages/dds/tree/src/shared-tree/transactionTypes.ts]: removed it.

File: packages/dds/tree/src/test/shared-tree/treeNodeApi.spec.ts (hunk 21 above)
Comment (agarwal-navin) [reply packages/dds/tree/src/test/shared-tree/treeNodeApi.spec.ts]: removed it.

File: packages/dds/tree/src/test/shared-tree/treeNodeApi.spec.ts (hunk 19 above)
Comment (agarwal-navin) [reply packages/dds/tree/src/test/shared-tree/treeNodeApi.spec.ts:378]: moved the tests.

File: packages/dds/tree/src/test/shared-tree/treeNodeApi.spec.ts (hunk 17 above)
Comment (agarwal-navin) [reply packages/dds/tree/src/test/shared-tree/treeNodeApi.spec.ts]: removed them.

File: packages/dds/tree/src/test/shared-tree/treeNodeApi.spec.ts (hunk 16 above)
Comment (agarwal-navin) [reply packages/dds/tree/src/test/shared-tree/treeNodeApi.spec.ts]: done

File: packages/dds/tree/src/shared-tree/schematizingTreeView.ts (hunk 11 above)
Comment (agarwal-navin) [reply packages/dds/tree/src/shared-tree/schematizingTreeView.ts]: Added doc comments.

File: packages/dds/tree/src/simple-tree/api/tree.ts (hunk 18 above)
Comment (agarwal-navin) [reply packages/dds/tree/src/simple-tree/api/tree.ts:581]: Added doc comments. Please take a look.

File: packages/dds/tree/src/shared-tree/schematizingTreeView.ts [hunk 23]
```diff
@@ -213,6 +223,68 @@ export class SchematizingSimpleTreeView<
 		return this.view;
 	}
 
+	/**
+	 * {@inheritDoc @fluidframework/shared-tree#TreeViewAlpha.runTransaction}
```
Comment (agarwal-navin) [inline packages/dds/tree/src/shared-tree/schematizingTreeView.ts:227]: @noencke @yann-achard-MS I am not sure if this will find the right doc comments. Do you guys know how I can tell?

File: packages/dds/tree/src/simple-tree/api/tree.ts [hunk 24]
```diff
@@ -542,6 +548,72 @@ export interface TreeViewAlpha<
...
+	 * - If reverted (e.g. via an "undo" operation), all the changes in the transaction are reverted together.
+	 * - The internal data representation of a transaction with many changes is generally smaller and more efficient than that of the changes when separate.
+	 *
+	 * Local change events will be emitted for each change as the transaction is being applied.
+	 * If the transaction is rolled back, a corresponding change event will also be emitted for the rollback.
+	 *
+	 * Nested transactions:
+	 * This API can be called from within the transaction callback of another runTransaction call. That will have slightly different behavior:
+	 * - If the inner transaction fails, only the inner transaction will be rolled back and the outer transaction will continue.
+	 * - Constraints will apply to the outermost transaction. Constraints are applied per commit and there will be one commit generated
```
Comment (agarwal-navin) [inline packages/dds/tree/src/simple-tree/api/tree.ts:577]: @noencke @yann-achard-MS This is the current behavior but it's not we discussed should happen going forward right. Basically, we want to throw if there are constraints in an inner transactions and undo will also behave differently.
Do you guys think I should do that here or it may be better to do it in a separate PR because I think that will require a discussion of its own.

File: packages/dds/tree/src/shared-tree/schematizingTreeView.ts (hunk 23 above)
Comment (noencke) [reply packages/dds/tree/src/shared-tree/schematizingTreeView.ts:227]: In my experience, @inheritDoc just doesn't work half the time, no matter how much you try to help it out. Sometimes it even works in the wrong way, and does things that are unexpected. I simply avoid using it.

File: packages/dds/tree/src/simple-tree/api/tree.ts (hunk 24 above)
Comment (noencke) [reply packages/dds/tree/src/simple-tree/api/tree.ts:577]: A separate PR is fine with me.

File: packages/dds/tree/src/simple-tree/api/tree.ts [hunk 25]
```diff
@@ -551,13 +551,65 @@ export interface TreeViewAlpha<
 
 	/**
 	 * Run a transaction which applies one or more edits to the tree as a single atomic unit.
-	 * @param transaction - The transaction callback to run.
-	 * @param params - The optional parameters for the transaction.
+	 * @param transaction - The function to run as the body of the transaction.
+	 * It should return a status object of {@link TransactionCallbackStatusExt | TransactionCallbackStatusExt } type.
+	 * It includes a "rollback" property which may be returned as "true" at any point during the transaction. This will
+	 * abort the transaction and discard any changes it made so far.
+	 * "rollback" can be set to true or left undefined to indicate that the body of the transaction has successfully run.
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/simple-tree/api/tree.ts:558]:
```suggestion
	 * "rollback" can be set to false or left undefined to indicate that the body of the transaction has successfully run.
```

File: packages/dds/tree/src/simple-tree/api/tree.ts [hunk 26]
```diff
@@ -551,13 +551,65 @@ export interface TreeViewAlpha<
...
+	 * @param params - The optional parameters for the transaction. It includes the constraints that will be checked before the transaction begins.
+	 * @returns A result object of {@link TransactionResultExt | TransactionResultExt} type. It includes the following:
+	 * - A "success" flag indicating whether the transaction was successful or not.
+	 * - The success of failure value as returned by the transaction function.
+	 * - This API will throw an error if the constraints are not met or something unexpected happens.
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/simple-tree/api/tree.ts:563]: This seems out of place in this list.

File: packages/dds/tree/src/shared-tree/transactionTypes.ts [hunk 27]
```diff
@@ -0,0 +1,116 @@
...
+ * @remarks The node must be in the document (its {@link TreeStatus | status} must be {@link TreeStatus.InDocument | InDocument}) to qualify as "existing".
+ * @public
+ */
+export interface NodeInDocumentConstraint {
+	readonly type: "nodeInDocument";
+	readonly node: TreeNode;
+}
+
+/**
+ * The status of a the transaction callback in the {@link RunTransaction | RunTransaction} API.
```
Comment (noencke) [inline packages/dds/tree/src/shared-tree/transactionTypes.ts:37]:
```suggestion
 * The status of the transaction callback in the {@link RunTransaction | RunTransaction} API.
```

File: packages/dds/tree/src/shared-tree/transactionTypes.ts [hunk 28]
```diff
@@ -0,0 +1,116 @@
...
+ * @alpha
+ */
+export type TransactionCallbackStatusExt<TSuccessValue, TFailureValue> =
```
Comment (noencke) [inline packages/dds/tree/src/shared-tree/transactionTypes.ts:40]:
```suggestion
export type TransactionCallbackStatus<TSuccessValue, TFailureValue> =
```

File: packages/dds/tree/src/shared-tree/transactionTypes.ts [hunk 29]
```diff
@@ -0,0 +1,116 @@
...
+			value: TFailureValue;
+	  };
+
+/**
+ * The status of a the transaction callback in the {@link RunTransaction | RunTransaction} API. This is the
+ * same as {@link TransactionCallbackStatusExt} but with the `value` field omitted. This is useful when
+ * the transaction callback doesn't need to return a value.
+ * @alpha
+ */
+export type TransactionCallbackStatus = Omit<
```
Comment (noencke) [inline packages/dds/tree/src/shared-tree/transactionTypes.ts:60]:
```suggestion
export type VoidTransactionCallbackStatus = Omit<
```

**Review by noencke — APPROVED** [review]

File: packages/dds/tree/src/simple-tree/api/tree.ts (hunk 26 above)
Comment (agarwal-navin) [reply packages/dds/tree/src/simple-tree/api/tree.ts:563]: Moved it to remarks.

**Review by yann-achard-MS — APPROVED** [review]
